README.md
# FVN

F is for friends

```
Swarm state: healthy

...2vYiga, 256.0 RPS  |                                 ########                             |
...vMfXzi, 3688.5 RPS |#########################                                             |
...i9CA9T, 2776.2 RPS |                              ###############                         |
...33Pgb9, 2108.6 RPS |###############                                                       |
...Ap3RUq, 256.0 RPS  |                                     ########                         |
...7mhsNH, 256.0 RPS  |                         ########                                     |
...iRUz6M, 2767.3 RPS |               ###############                                        |
...zbDrjo, 3500.3 RPS |                                             #########################|


Legend:

# - online
J - joining     (loading blocks)
? - unreachable (port forwarding/NAT/firewall issues, see below)
_ - offline     (just disconnected)
```

## Features

This extension will do following:

1. Connect to a swarm of [GUN](https://gun.eco) bootstrap peers.
2. Provide a query interface via the `Fire` command (see below).
3. Host a shared API at: https://localhost:60666/gun
4. If multiple instances of VSCode are running, the extension will connect to a single, local API. In the event of failure, VSCode will attempt to re-launch the API.

## Usage

This extension exports a single command, called "Fire". To fire a bullet into GUN (making a query), use these steps:

1. Press `Ctrl + Shift + P`
2. Type `Fire`

To automatically scroll with the extension, click on the "lock" icon in the upper-right corner of the Output tab.

More features to come...

## Installation Methods

### 1. From the offline, packaged VSCode extension

Browse the [./release](./release) folder, and install the latest .vsix file via:

```
Extensions > ... > Install from VSIX...
```

### 2. Install from source

Clone this repository, and open it in VSCode. Use `npm install` to load dependencies. Press "F5" to start the debugger.

If all goes well, a new VSCode window will open, running this extension.

## Extension Settings

Include if your extension adds any VS Code settings through the `contributes.configuration` extension point.

For example:

This extension contributes the following settings:

-   `myExtension.enable`: Enable/disable this extension.
-   `myExtension.thing`: Set to `blah` to do something.


CHANGELOG.md
# Change Log

All notable changes to the "fvn" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## Release Notes

### 0.1.0

-   Initial release...


src/gun.js
const http = require('http')
const Gun = require('gun')
const {
    colors,
    delay,
    isAlive,
    outputChannel,
    randomString
} = require('./common')

class FVN {
    constructor() {
        this.localHost = 'STATE'
        this.localPort = 60666
        this.localPeer = `http://localhost:${this.localPort}/gun`
        this.bootstrapPeers = [
            'https://59.src.eco/gun',
            'https://95.src.eco/gun'
        ]
        this.currentPeers = []
        this.keepAliveCounter = 3
        this.focus = 'trade'
        this.identifier = randomString(18, '0123456789')
        this.pubKey = null
        this.gun = null
        this.previousBullet = null
        this.init()
    }

    async init() {
        if (!(await isAlive(this.localPeer))) {
            outputChannel.appendLine('GUN is offline, starting server...')
            this.startServer()
        } else {
            this.currentPeers = [this.localPeer]
            this.connectGun()
        }
    }

    async keepAlive() {
        await delay(15000)
        const peers = this.gun.back('opt.peers')
        for (const i of this.currentPeers) {
            const state = peers[i]?.wire?.readyState
            if (state === 0 || state === null || typeof state === 'undefined') {
                outputChannel.appendLine(
                    `Local GUN API is offline. Will retry ${this.keepAliveCounter.toString()} times before hosting it myself...`
                )
                this.keepAliveCounter--
                this.gun.opt({ peers: [...this.currentPeers] })
                if (this.keepAliveCounter === 0) {
                    this.currentPeers = this.bootstrapPeers
                    this.startServer(false)
                }
            } else {
                this.keepAliveCounter = 3
            }
        }
        this.keepAlive()
    }

    startServer(init = true) {
        const requestListener = (req, res) => {
            res.writeHead(200)
            res.end(`${this.previousBullet}`)
        }

        const server = http
            .createServer(requestListener)
            .listen(60666, '0.0.0.0', () => {
                const message = `GUN is listening at: ${this.localPeer}`
                console.log(message)
                outputChannel.appendLine(message)
            })

        this.currentPeers = this.bootstrapPeers
        this.connectGun(server, init)
    }

    connectGun(server = null, init = true) {
        this.gun = Gun({
            peers: this.currentPeers,
            web: server,
            file: './gun',
            localStorage: false,
            radisk: false,
            axe: false
        })
        // Required for keepAlive()
        this.gun.get('src').on((data) => {})
        this.logMessages()
        if (init) {
            this.keepAlive()
        }
    }

    sendMessage(input) {
        const bullet = JSON.stringify({
            focus: this.focus,
            message: input,
            pubKey: this.pubKey,
            identifier: this.identifier,
            mode: 'cos'
        })
        this.previousBullet = bullet
        this.gun.get('src').get('bullets').get(this.focus).put(bullet)
        console.log(
            `${colors.RED}< ONE@${this.localHost}:${colors.WHITE} ${input}`
        )
        outputChannel.appendLine(`< ONE@${this.localHost}: ${input}`)
    }

    logMessages() {
        this.gun
            .get('src')
            .get('bullets')
            .get(this.focus)
            .on(async (bullet) => {
                if (this.previousBullet === bullet) return
                this.previousBullet = bullet
                const data = JSON.parse(bullet)
                let message = data.message
                let color = colors.RED
                if (
                    data.pubKey !== null &&
                    typeof data.pubKey !== 'undefined'
                ) {
                    const sender = await this.gun.user(data.pubKey)
                    if (typeof sender === 'undefined') {
                        color = colors.BLUE
                    } else
                        message = await Gun.SEA.verify(data.message, sender.pub)
                    color = colors.PURPLE
                } else {
                    color = colors.BLUE
                }
                console.log(
                    `${color}> ONE@${this.localHost}:${colors.WHITE} ${message}`
                )
                outputChannel.appendLine(`> ONE@${this.localHost}: ${message}`)
            })
    }
}

module.exports = FVN


src/common.js
const vscode = require('vscode')
const http = require('http')

const colors = {
    BLUE: '\x1b[34m',
    GREEN: '\x1b[32m',
    PURPLE: '\x1b[95m',
    RED: '\x1b[31m',
    WHITE: '\x1b[0m'
}

const delay = (ms) => new Promise((res) => setTimeout(res, ms))

function randomString(
    len,
    chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
) {
    let text = ''
    for (let i = 0; i < len; i++) {
        text += chars.charAt(Math.floor(Math.random() * chars.length))
    }
    return text
}

const outputChannel = vscode.window.createOutputChannel('FVN', 'src')

async function isAlive(url) {
    return new Promise((resolve, reject) => {
        const request = http.get(url, (response) => {
            if (response.statusCode === 200) {
                return resolve(true)
            } else {
                resolve(false)
            }
        })

        request.on('error', (err) => {
            console.error(err)
            resolve(false)
        })

        request.end()
    })
}

function sigmoid(x) {
    return 1 / (1 + Math.exp(-x))
}

module.exports = {
    colors,
    delay,
    isAlive,
    outputChannel,
    randomString,
    sigmoid
}


src/language.js
const vscode = require('vscode')
const tokenTypes = [
    'class',
    'interface',
    'enum',
    'function',
    'variable',
    'keyword'
]
const tokenModifiers = ['declaration', 'documentation', 'input']
const legend = new vscode.SemanticTokensLegend(tokenTypes, tokenModifiers)

const provider = {
    provideDocumentSemanticTokens(document) {
        // analyze the document and return semantic tokens

        const tokensBuilder = new vscode.SemanticTokensBuilder(legend)
        // on line 1, characters 1-5 are a class declaration

        tokensBuilder.push(
            new vscode.Range(
                new vscode.Position(1, 0),
                new vscode.Position(1, 11)
            ),
            'keyword',
            ['input']
        )

        return tokensBuilder.build()
    }
}

const selector = { language: 'src' } // register for all Java documents from the local file system

vscode.languages.registerDocumentSemanticTokensProvider(
    selector,
    provider,
    legend
)


src/extension.js
const vscode = require('vscode')
const { outputChannel } = require('./common')
const FVN = require('./fvn')

/**
 * This function is called when the extension is activated
 * @param {vscode.ExtensionContext} context
 */
async function activate(context) {
    context.subscriptions.push(outputChannel)

    outputChannel.show(true)

    const fvn = new FVN()

    let disposable = vscode.commands.registerCommand('fvn.fire', function () {
        outputChannel.show(true)
        vscode.window
            .showInputBox({ prompt: 'Send message:' })
            .then(async (input) => {
                if (input) {
                    fvn.sendMessage(input)
                }
            })
    })

    context.subscriptions.push(disposable)

    vscode.window.setStatusBarMessage('GUN is online. Have FVN!')
}

// This method is called when the extension is deactivated
function deactivate() {}

module.exports = {
    activate,
    deactivate
}


src/fvn.js
const vscode = require('vscode')
const http = require('http')
const Gun = require('gun')
const {
    colors,
    delay,
    isAlive,
    outputChannel,
    randomString
} = require('./common')

class FVN {
    constructor() {
        this.localHost = 'STATE'
        this.localPort = 60666
        this.localPeer = `http://localhost:${this.localPort}/gun`
        this.bootstrapPeers = [
            'https://59.src.eco/gun',
            'https://95.src.eco/gun'
        ]
        this.currentPeers = []
        this.keepAliveCounter = 3
        this.focus = 'trade'
        this.identifier = randomString(18, '0123456789')
        this.pubKey = null
        this.gun = null
        this.previousBullet = null
        this.init()
    }

    async init() {
        if (!(await isAlive(this.localPeer))) {
            outputChannel.appendLine('GUN is offline, starting server...')
            this.startServer()
        } else {
            this.currentPeers = [this.localPeer]
            this.connectGun()
        }
    }

    async keepAlive() {
        await delay(15000)
        const peers = this.gun.back('opt.peers')
        for (const i of this.currentPeers) {
            const state = peers[i]?.wire?.readyState
            if (state === 0 || state === null || typeof state === 'undefined') {
                outputChannel.appendLine(
                    `Local GUN API is offline. Will retry ${this.keepAliveCounter.toString()} times before hosting it myself...`
                )
                this.keepAliveCounter--
                this.gun.opt({ peers: [...this.currentPeers] })
                if (this.keepAliveCounter === 0) {
                    this.currentPeers = this.bootstrapPeers
                    this.startServer(false)
                }
            } else {
                this.keepAliveCounter = 3
            }
        }
        this.keepAlive()
    }

    startServer(init = true) {
        const requestListener = (req, res) => {
            res.writeHead(200)
            res.end(`${this.previousBullet}`)
        }

        const server = http
            .createServer(requestListener)
            .listen(60666, '0.0.0.0', () => {
                const message = `GUN is listening at: ${this.localPeer}`
                console.log(message)
                outputChannel.appendLine(message)
            })

        this.currentPeers = this.bootstrapPeers
        this.connectGun(server, init)
    }

    connectGun(server = null, init = true) {
        this.gun = Gun({
            peers: this.currentPeers,
            web: server,
            file: './gun',
            localStorage: false,
            radisk: false,
            axe: false
        })
        // Required for keepAlive()
        this.gun.get('src').on((data) => {})
        this.logMessages()
        if (init) {
            this.keepAlive()
        }
    }

    sendMessage(input) {
        const bullet = JSON.stringify({
            focus: this.focus,
            message: input,
            pubKey: this.pubKey,
            identifier: this.identifier,
            mode: 'cos'
        })
        this.previousBullet = bullet
        this.gun.get('src').get('bullets').get(this.focus).put(bullet)
        console.log(
            `${colors.RED}< ONE@${this.localHost}:${colors.WHITE} ${input}`
        )
        outputChannel.appendLine(`< ONE@${this.localHost}: ${input}`)
    }

    logMessages() {
        this.gun
            .get('src')
            .get('bullets')
            .get(this.focus)
            .on(async (bullet) => {
                if (this.previousBullet === bullet) return
                this.previousBullet = bullet
                const data = JSON.parse(bullet)
                let message = data.message
                let color = colors.RED
                if (
                    data.pubKey !== null &&
                    typeof data.pubKey !== 'undefined'
                ) {
                    const sender = await this.gun.user(data.pubKey)
                    if (typeof sender === 'undefined') {
                        color = colors.BLUE
                    } else
                        message = await Gun.SEA.verify(data.message, sender.pub)
                    color = colors.PURPLE
                } else {
                    color = colors.BLUE
                }
                const result = await simulator(message)
                console.log(
                    `${color}> ONE@${this.localHost}:${colors.WHITE} ${result.message}`
                )
                outputChannel.appendLine(
                    `> ONE@${this.localHost}: ${result.message}`
                )
                if (result.choice1 && result.choice2) {
                    const response = await vscode.window.showInformationMessage(
                        'Someone has requested your action.',
                        'GoTo'
                    )
                    if (response === 'GoTo') {
                        outputChannel.show(true)
                        const choice =
                            await vscode.window.showInformationMessage(
                                'Simulator: Make a choice.',
                                result.choice1,
                                result.choice2
                            )

                        outputChannel.appendLine(
                            `< ONE@${this.localHost}: ${choice}`
                        )
                    }
                }
            })
    }
}

// This doesn't really do anything, except for demonstrating the workflow.
async function simulator(message) {
    if (Math.random() < 0.001) {
        const words = message.split(' ')
        if (words.length < 3) return { message }
        const randomIndexes = []
        while (randomIndexes.length < 2) {
            const randomIndex = Math.floor(Math.random() * words.length)
            if (!randomIndexes.includes(randomIndex)) {
                randomIndexes.push(randomIndex)
                words[randomIndex] = `((${words[randomIndex]}))`
            }
        }
        message = words.join(' ')
        return {
            message,
            choice1: words[randomIndexes[0]],
            choice2: words[randomIndexes[1]]
        }
    }
    return { message }
}

module.exports = FVN
